pkg/controller/jobs/kubeflow/trainjob/trainjob_controller.go

```diff
@@ -0,0 +1,291 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trainjob
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+
+	kftrainerapi "github.com/kubeflow/trainer/v2/pkg/apis/trainer/v1alpha1"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	jobsetapi "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+	kJobset "sigs.k8s.io/kueue/pkg/controller/jobs/jobset"
+	"sigs.k8s.io/kueue/pkg/features"
+	"sigs.k8s.io/kueue/pkg/podset"
+	clientutil "sigs.k8s.io/kueue/pkg/util/client"
+)
+
+var (
+	gvk                    = kftrainerapi.GroupVersion.WithKind("TrainJob")
+	FrameworkName          = "trainer.kubeflow.org/trainjob"
+	TrainJobControllerName = "trainer.kubeflow.org/trainjob-controller"
+)
+
+const (
+	FirstOverrideIdx = "kueue.x-k8s.io/override-idx"
+)
+
+func init() {
+	utilruntime.Must(jobframework.RegisterIntegration(FrameworkName, jobframework.IntegrationCallbacks{
+		SetupIndexes:      SetupIndexes,
+		NewJob:            NewJob,
+		NewReconciler:     NewReconciler,
+		SetupWebhook:      SetupTrainJobWebhook,
+		JobType:           &kftrainerapi.TrainJob{},
+		AddToScheme:       kftrainerapi.AddToScheme,
+		MultiKueueAdapter: &multiKueueAdapter{},
+	}))
+}
+
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainjobs,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainjobs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainjobs/finalizers,verbs=get;update
+
+type trainJobReconciler struct {
+	jr     *jobframework.JobReconciler
+	client client.Client
+}
+
+var reconciler trainJobReconciler
+var _ jobframework.JobReconcilerInterface = (*trainJobReconciler)(nil)
+
+func NewReconciler(client client.Client, eventRecorder record.EventRecorder, opts ...jobframework.Option) jobframework.JobReconcilerInterface {
+	reconciler = trainJobReconciler{
+		jr:     jobframework.NewReconciler(client, eventRecorder, opts...),
+		client: client,
+	}
+	return &reconciler
+}
+
+func (r *trainJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
I'm wonder if we can use Skip() instead.

https://github.com/kaisoz/kueue/blob/2c5839f1ecbbf6f30f5db1f8130f02f8123349dc/pkg/controller/jobframework/interface.go#L104-L106

Reply:
Can you clarify what type of optimization, is "performance" or code "cleanliness"? If performance, the what is the expected gain. 

Also, I'm not clear how you suggest to implement the Skip, could you clarify?

Reply:
Yes, sure, it's just code cleanup. 

Instead of this:
```
func (r *trainJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
	jobSet := jobsetapi.JobSet{}
	err := r.client.Get(ctx, req.NamespacedName, &jobSet)
	if err != nil {
		// we only reconcile the TrainJob when its child JobSet is available
		return ctrl.Result{}, client.IgnoreNotFound(err)
	}
	return r.jr.ReconcileGenericJob(ctx, req, &TrainJob{})
}
```

I propose to move logic:

```
	jobSet := jobsetapi.JobSet{}
	err := r.client.Get(ctx, req.NamespacedName, &jobSet)
	if err != nil {
		// we only reconcile the TrainJob when its child JobSet is available
		return ctrl.Result{}, client.IgnoreNotFound(err)
	}
```

to 

Skip() bool

Reply:
thx for the comment @mbobrovskyi . Anyway, I'm working on a follow-up to avoid having to watch the `Jobset` so that piece of code will disappear